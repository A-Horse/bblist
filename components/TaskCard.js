import React, {Component} from 'react';
import {connect} from 'react-redux';
import {browserHistory} from 'react-router';
import Radium from 'radium';
import R from 'fw-ramda';

import {deleteTaskWall, getTaskAllCards} from '../actions/task-wall';
import {postTaskCard} from '../actions/task-card';
import {createTaskList, deleteTaskList} from '../actions/task-list';
import {DropMenu} from './widget/DropMenu';
import {ConfirmModal} from './widget/ConfirmModal';
import {getAssets} from '../services/assets-manager';
import {spawnThemeRender} from '../style/theme-render';
import {PageContainer} from './widget/PageContainer';
import {AddIcon, EditIcon, ArrowDownIcon, SettingIcon} from '../services/svg-icons';
import {TaskWallSetting} from './TaskWallSetting';
import {navHeight} from './Nav';
import {Hr} from './widget/Hr';

const styles = {
  card: {
    margin: '0.2rem 0',
    padding: '4px 8px',
    borderRadius: '1px',
    height: '48px'
  }
};

const themeRender = spawnThemeRender(styles);
themeRender('card', 'lightBackground', 'lightSmallShadow');

@Radium
class TaskCard extends Component {
  constructor() {
    super();
  }

  componentWillMount() {
    this.state = {
      
    };
  }

  

  render() {
    const {card} = this.props;
    return (
        <div style={styles.card}>
          <p>{card.title}</p>     
        </div>
    );
  }

}

const mapStateToProps = (state) => {
  return {
    wallData: state.taskWall.wallData
  };
}

export default connect(mapStateToProps)(TaskCard);
