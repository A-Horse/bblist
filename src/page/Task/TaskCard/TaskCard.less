.task-card {
  flex: 1 0 auto;
  background: white;
  white-space: normal;
  word-break: break-all;
  display: flex;
  justify-content: space-between;
  align-items: center;
  margin: 4px 0;
  padding: 12px;
  box-shadow: 0 2px 4px #eee;
  cursor: pointer;

  &.long {
    border-radius: 3px;
    padding: 5px 6px;
    margin: 6px;

    .useravatar {
      width: 25px;
      height: 25px;
    }
  }

  .task-card--title {
    padding: 0 12px;
    width: 205px;
  }

  .useravatar {
    width: 39px;
  }

  &:first-child {
    margin-top: 0px;
  }
}


.task-card-placeholder {
  width: 100%;
  background: #f8f8f8;
  box-shadow: 0 2px 4px #eee;
  margin: 4px 0;
}
