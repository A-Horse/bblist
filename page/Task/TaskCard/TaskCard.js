import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { openTaskCardModal } from 'actions/event/task-wall';
import UserAvatar from 'components/UserAvatar';
import { CheckBox } from 'components/widget/CheckBox/CheckBox';
import { getMouseElementInnerOffset } from 'utils/dom';
import { browserHistory } from 'react-router';

import './TaskCard.scss';

class TaskCard extends Component {
  static propTypes = {
    card: PropTypes.object.isRequired
  };

  state = {};

  constructor(props) {
    super(props);
    this.updateCard = this.updateCard.bind(this);
    this.onLoad = this.onLoad.bind(this);
  }

  componentWillMount() {}

  onDragEnd() {
    document.body.removeChild(this.crt);
  }

  openTaskCardModal() {
    return dispatch(openTaskCardModal(this.props.card));
  }

  /* updateDone() {
   *   const isDone = this.refs.checkbox.checked;
   *   this.updateTaskCard({ isDone });
   * }*/

  updateCard(toPatchData) {
    /* const { dispatch } = this.props;
     * return dispatch(updateTaskCard(this.props.cardId, data));*/
    this.props.actions.UPDATE_TASK_CARD_REQUEST({
      id: this.props.card.get('id'),
      ...toPatchData
    });
  }

  onLoad() {
    this.height = this.refs.main.offsetHeight;
    this.width = this.refs.main.offsetWidth;
  }

  onClick(event) {
    browserHistory.push(`/task-board/${this.props.boardId}/${this.props.cardId}`);
  }

  checkBoxOnClick(event) {
    event.stopPropagation();
  }

  onMouseDown(event) {
    // TODO  尝试一下 react 的方法
    this.mouseMoving = false;

    event.preventDefault();
    event.stopPropagation();
    const tracks = window.document.querySelectorAll('.task-track');

    const thisCard = this.refs.main;

    const movingCard = thisCard.cloneNode(true);
    const pageContainer = window.document.body.querySelector('.board-page-container');

    const mouseOffsetInCard = getMouseElementInnerOffset(thisCard, event);

    const self = this;
    function onMouseMove(event) {
      if (!self.mouseMoving) {
        // 防止点击的时候还是出现 clone
        movingCard.style.height = thisCard.offsetHeight;
        movingCard.style.width = thisCard.offsetWidth;
        movingCard.style.position = 'absolute';
        movingCard.style.left = event.pageX + 'px';
        movingCard.style.top = event.pageY + 'px';

        window.document.body.appendChild(movingCard);
      }
      self.mouseMoving = true;
      const movingOffsetY = event.pageY;
      const movingOffsetX = event.pageX + pageContainer.scrollLeft;
      movingCard.style.left = '0';
      movingCard.style.top = '0';
      movingCard.style.transform = `translate(${movingOffsetX -
        mouseOffsetInCard.left}px, ${movingOffsetY - mouseOffsetInCard.top}px)`;
    }

    function onMouseUp(event) {
      window.document.body.removeEventListener('mousemove', onMouseMove);
      window.document.body.removeEventListener('mouseup', onMouseUp);
      if (self.mouseMoving) {
        window.document.body.removeChild(movingCard);
      }
    }
    window.document.body.addEventListener('mousemove', onMouseMove);
    window.document.body.addEventListener('mouseup', onMouseUp);
  }

  render() {
    const { card } = this.props;
    return (
      <div className="task-card" ref="main">
        <CheckBox
          ref="checkbox"
          defaultChecked={card.get('isDone')}
          onChange={checked => this.updateCard({ isDone: checked })}
        />
        <p className="task-card--title">{card.get('title')}</p>
        <UserAvatar user={card.get('creater').toJS()} />
      </div>
    );
  }
}

export default TaskCard;
