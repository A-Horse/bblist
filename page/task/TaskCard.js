import React, {Component} from 'react';
import {connect} from 'react-redux';

import {deleteTaskWall, getTaskAllCards} from 'actions/task/task-wall';
import {updateTaskCard, insertVirtualCard, activeCardModal} from 'actions/task/task-card';
import {createTaskList, deleteTaskList} from 'actions/task/task-list';
import {openTaskCardModal} from 'actions/event/task-wall';
import {DropList} from 'components/widget/DropList';
import {ConfirmModal} from 'components/widget/ConfirmModal';
import {PageContainer} from 'components/widget/PageContainer';
import {Hr} from 'components/widget/Hr';
import {getAssets} from 'services/assets-manager';
import {spawnMixinRender} from 'style/theme-render';
import {AddIcon, EditIcon, ArrowDownIcon, SettingIcon} from 'services/svg-icons';
import {navHeight} from 'components/Nav';
import UserAvatar from 'components/UserAvatar';
import {CheckBox} from 'components/widget/CheckBox';
import BoardCradDragHelper from 'services/board-card-drag-helper';
import {getMouseElementInnerOffset} from 'utils/dom';

import 'style/page/task/card.scss';

class TaskCard extends Component {
  constructor() {
    super();
    this.state = {};
  }

  componentWillMount() {
  }

  // onDragStart(event) {
  //   // TODO 移动的图标
  //   const {normalizedCard, cardId} = this.props;
  //   const card = normalizedCard.entities[cardId];
    
  //   const crt = this.refs.main.cloneNode(true);

  //   const width = this.refs.main.offsetWidth;
  //   const height = this.refs.main.offsetHeight;

  //   // TODO extract function
  //   this.crt = crt;

  //   crt.style.position = 'absolute';
  //   crt.style.top = '-100%';
  //   crt.style.right = '-100%';
  //   crt.style.height = height + 'px';
  //   crt.style.width = width + 'px';
  //   document.body.appendChild(crt);
  //   event.dataTransfer.setDragImage(crt, 0, 0);
    
  //   BoardCradDragHelper.setData('info', {
  //     from: {
  //       listId: card.taskListId
  //     },
  //     offsetX: event.nativeEvent.offsetX,
  //     offsetY: event.nativeEvent.offsetY,
  //     width,
  //     height
  //   });
    
  //   this.refs.main.style.position = 'absolute';
  //   this.isDragging = true;
  // }
  
  onDragEnd(event) {
    document.body.removeChild(this.crt);
  }
  
  openTaskCardModal() {
    const {dispatch} = this.props;
    return dispatch(openTaskCardModal(this.props.card));
  }

  updateDone() {
    const isDone = this.refs.checkbox.checked;
    this.updateTaskCard({isDone});    
  }

  updateTaskCard(data) {
    const {dispatch} = this.props;
    return dispatch(updateTaskCard(this.props.card.id, data));
  }

  onLoad() {
    this.height = this.refs.main.offsetHeight;
    this.width = this.refs.main.offsetWidth;
  }

  onClick(event) {
    const {dispatch} = this.props;
    return dispatch(activeCardModal(this.props.cardId));
  }

  checkBoxOnClick(event) {
    event.stopPropagation();
  }

  onMouseDown(event) {
    // TODO  尝试一下 react 的方法
    this.mouseMoving = false;

    event.preventDefault();
    event.stopPropagation();
    const tracks = window.document.querySelectorAll('.task-list');

    const thisCard = this.refs.main;
    
    const movingCard = thisCard.cloneNode(true);
    const pageContainer = window.document.body.querySelector('.board-page-container');

    const mouseOffsetInCard = getMouseElementInnerOffset(thisCard, event);

    const self = this;
    function onMouseMove(event) {
      if (!self.mouseMoving) {
        movingCard.style.height = thisCard.offsetHeight;
        movingCard.style.width = thisCard.offsetWidth;
        movingCard.style.position = 'absolute';
        movingCard.style.left = event.pageX + 'px';
        movingCard.style.top = event.pageY + 'px';
        
        window.document.body.appendChild(movingCard);
      }
      self.mouseMoving = true;
      const movingOffsetY = event.pageY;
      const movingOffsetX = event.pageX + pageContainer.scrollLeft;
      movingCard.style.left = '0';
      movingCard.style.top = '0';
      movingCard.style.transform = `translate(${movingOffsetX - mouseOffsetInCard.left}px, ${movingOffsetY - mouseOffsetInCard.top}px)`;
    }
    
    function onMouseUp(event) {
      window.document.body.removeChild(movingCard);
      window.document.body.removeEventListener('mousemove', onMouseMove);
      window.document.body.removeEventListener('mouseup', onMouseUp);
    }

    window.document.body.addEventListener('mousemove', onMouseMove);
    window.document.body.addEventListener('mouseup', onMouseUp);
  }

  render() {
    const {normalizedCard, cardId} = this.props;
    const card = normalizedCard.entities[cardId];
    const activeRole = card.creater;
    return (
      <div className='task-card' ref='main'
           onMouseDown={this.onMouseDown.bind(this)}
           onClick={this.onClick.bind(this)}
           onLoad={this.onLoad.bind(this)}>
        <CheckBox ref='checkbox' defaultChecked={card.isDone} onChange={this.updateDone.bind(this)} onClick={this.checkBoxOnClick.bind(this)}/>
        <p className='task-card--title'>{card.title}</p>
        <UserAvatar user={activeRole}/>
      </div>
    );
  }
}

const mapStateToProps = (state) => {
  return {
    normalizedCard: state.task.card,
    normalizedTrack: state.task.list
  };
};

export default connect(mapStateToProps, null, null, {withRef: true})(TaskCard);
