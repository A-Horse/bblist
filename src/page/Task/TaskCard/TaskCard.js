// @flow
import React, { Component } from 'react';
import { UserAvatar } from '../../../components/UserAvatar/UserAvatar';
import { Checkbox } from 'antd';
import { getMouseElementInnerOffset } from '../../../utils/dom';
import { DragSource, DropTarget, ConnectDragSource } from 'react-dnd';
import { findDOMNode } from 'react-dom';
import { Map } from 'immutable';

import './TaskCard.less';

const TaskCardSource = {
  beginDrag(props, monitor, component) {
    const componentRect = findDOMNode(component).getBoundingClientRect();
    return {
      card: props.card,
      height: componentRect.height
    };
  }
};

@DropTarget(
  'CARD',
  {
    drop(props, monitor, component) {
      console.log(component);
      /* component.onTaskCardDrop(props.card); */
      props.actions.CARD_MOVE_REQUEST({
        sourceCard: monitor.getItem().card.toJS(),
        targetCard: props.card.toJS()
      });
    },
    canDrop(props, monitor) {
      const item = monitor.getItem();
      return item.card.get('id') !== props.card.get('id');
    }
  },
  (connect, monitor) => ({
    connectDropTarget: connect.dropTarget(),
    isOver: monitor.isOver(),
    isOverCurrent: monitor.isOver({ shallow: true }),
    canDrop: monitor.canDrop(),
    itemType: monitor.getItemType(),
    sourceItem: monitor.getItem()
  })
)
@DragSource('CARD', TaskCardSource, (connect, monitor) => ({
  connectDragSource: connect.dragSource(),
  connectDragPreview: connect.dragPreview(),
  isDragging: monitor.isDragging()
}))
export class TaskCard extends Component<
  {
    card: any,
    actions: any,
    history: any,
    match: any
  },
  {}
> {
  state = {};

  componentWillMount() {}

  componentWillReceiveProps(newProps) {}

  updateCard = (toPatchData: any) => {
    this.props.actions.UPDATE_TASK_CARD_REQUEST({
      id: this.props.card.get('id'),
      ...toPatchData
    });
  };

  render() {
    const { card } = this.props;
    const { isDragging, connectDragSource, connectDragPreview } = this.props;
    const { isOver, canDrop, connectDropTarget } = this.props;
    const opacity = isDragging ? 0.4 : 1;
    return (
      connectDragPreview &&
      connectDragSource &&
      connectDragPreview(
        connectDropTarget(
          <div style={{ opacity }}>
            {connectDragSource(
              <div>
                <div className="task-card">
                  {card.get('id')}
                  <Checkbox
                    checked={card.get('isDone')}
                    onChange={event => this.updateCard({ isDone: event.target.checked })}
                  />
                  <p
                    className="task-card--title"
                    onClick={() => {
                      this.props.history.push(this.props.match.url + `/card/${card.get('id')}`);
                    }}
                  >
                    {card.get('title')}
                  </p>
                  <UserAvatar user={card.get('creater').toJS()} />
                </div>
                <div
                  style={{
                    height:
                      this.props.sourceItem && this.props.isOver && this.props.canDrop
                        ? this.props.sourceItem.height
                        : 0
                  }}
                  className={`task-card-placeholder${
                    this.props.isOver && this.props.canDrop ? ' active' : ''
                  }`}
                />
              </div>
            )}
          </div>
        )
      )
    );
  }
}
