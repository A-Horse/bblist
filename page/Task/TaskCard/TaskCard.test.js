import React from 'react';
import TaskCard from './TaskCard';
import renderer from 'react-test-renderer';
import { fromJS } from 'immutable';

test('TaskCard', () => {
  const card = fromJS({
    id: 23,
    title: 'sdsd!!!',
    taskListId: 17,
    sprint: null,
    createrId: 2,
    ownerId: null,
    executorId: null,
    id: 23,
    index: 4,
    isDone: 1,
    owner: {},
    ownerId: null,
    sprint: null,
    status: null,
    taskListId: 17,
    taskWallId: 15,
    title: 'taskcard',
    updated_at: null,
    creater: {
      name: 'God',
      email: 'god@ocotpus.com',
      id: 0
    }
  });

  const tree = renderer.create(<TaskCard card={card} />);
  expect(tree).toMatchSnapshot();
});
