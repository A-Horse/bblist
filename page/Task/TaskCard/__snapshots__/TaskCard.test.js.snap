// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskCard 1`] = `
<div
  className="task-card"
>
  <div
    className="checkbox"
    onClick={undefined}
  >
    <label
      className="checkbox--mock"
    >
      <input
        defaultChecked={1}
        onChange={[Function]}
        type="checkbox"
      />
      <i
        aria-hidden="true"
        className="fa fa-check-square-o"
      />
    </label>
  </div>
  <p
    className="task-card--title"
  >
    taskcard
  </p>
  <img
    className="useravatar"
    crossOrigin="Anonymous"
    src="https://www.gravatar.com/avatar/0f6dbc559e6c30c7a6294a7bcb62cebd"
    style={undefined}
  />
</div>
`;
